server/src/main/java/org/opensearch/action/bulk/TransportBulkAction.java

```diff
@@ -595,14 +603,28 @@ protected void doRun() {
             }
 
             if (requestsByShard.isEmpty()) {
-                listener.onResponse(
-                    new BulkResponse(responses.toArray(new BulkItemResponse[responses.length()]), buildTookInMillis(startTimeNanos))
-                );
+
+                BulkItemResponse[] response = responses.toArray(new BulkItemResponse[responses.length()]);
+                long tookMillis = buildTookInMillis(startTimeNanos);
+
+                for (BulkItemResponse itemResponse : response) {
+                    if (itemResponse != null) {
+                        indicesService.incDocStatusCounter(itemResponse.status());
+                    }
+                }
+
+                listener.onResponse(new BulkResponse(response, tookMillis));
```

Comment:
Instead of inline instrumentation, can we add a wrapper listener like BulkStatsAwareActionListener which collects stats on onResponse and onFailure

Reply:
It's done for performance benefits.